ampd/src/stellar/rpc_client.rs

```
@@ -28,13 +28,22 @@ const STATUS_SUCCESS: &str = "SUCCESS";
 
 impl From<(Hash, GetTransactionResponse)> for TxResponse {
     fn from((transaction_hash, response): (Hash, GetTransactionResponse)) -> Self {
-        let contract_events = response
-            .result_meta
-            .and_then(|meta| match meta {
-                TransactionMeta::V3(data) => data.soroban_meta.map(|meta| meta.events),
-                _ => None,
-            })
-            .unwrap_or_default();
+        // Protocol 23 (CAP-0067): Extract contract events from the unified events structure
+        // contract_events is Vec<Vec<ContractEvent>> (per operation), so we flatten it
+        let events_vec: Vec<ContractEvent> = response
+            .events
+            .contract_events
+            .into_iter()
+            .flatten()
+            .collect();
+
+        let contract_events = events_vec.try_into().unwrap_or_else(|_| {
+            warn!(
+                tx_hash = %transaction_hash,
+                "Contract events exceed VecM capacity, returning empty list"
+            );
+            VecM::default()
```

hmm, silent failure. we should handle the failure explicitly, probably something we can do with the ampd handler approach better

---

I see. I added a logic to handle the failure explicitly. please take a look

---

can we just use a regular Vec here? I don't see why we need to use this `VecM` type. Also, it looks like the capacity is u32::Max, doubt we will ever hit this case.